indispensable/src/commonMain/kotlin/at/asitplus/signum/indispensable/asn1/encoding/NumberEncoding.kt

```diff
@@ -241,222 +163,374 @@ fun Int.toUnsignedByteArray() = toLong().toUnsignedByteArray()
  * Encodes this number using varint encoding as used within ASN.1: groups of seven bits are encoded into a byte,
  * while the highest bit indicates if more bytes are to come
  */
-fun ULong.toAsn1VarInt(): ByteArray {
-    if (this < 128u) return byteArrayOf(this.toByte()) //Fast case
-    var offset = 0
-    var result = mutableListOf<Byte>()
+fun ULong.toAsn1VarInt(): ByteArray = throughBuffer { it.writeAsn1VarInt(this) }
 
-    var b0 = (this shr offset and 0x7FuL).toByte()
-    while ((this shr offset > 0uL) || offset == 0) {
-        result += b0
-        offset += 7
-        if (offset > (ULong.SIZE_BITS - 1)) break //End of Fahnenstange
-        b0 = (this shr offset and 0x7FuL).toByte()
-    }
-    return with(result) {
-        ByteArray(size) { fromBack(it) or asn1VarIntByteMask(it) }
-    }
-}
+/**
+ * Encodes this number using varint encoding as used within ASN.1: groups of seven bits are encoded into a byte,
+ * while the highest bit indicates if more bytes are to come
+ */
+@Throws(IllegalArgumentException::class)
+fun BigInteger.toAsn1VarInt(): ByteArray = throughBuffer { it.writeAsn1VarInt(this) }
+
+/**
+ * Encodes this number using unsigned VarInt encoding as used within ASN.1:
+ * Groups of seven bits are encoded into a byte, while the highest bit indicates if more bytes are to come.
+ *
+ * This kind of encoding is used to encode [ObjectIdentifier] nodes and ASN.1 Tag values > 30
+ */
+fun UInt.toAsn1VarInt(): ByteArray = throughBuffer { it.writeAsn1VarInt(this) }
+
+private fun MutableList<Byte>.asn1VarIntByteMask(it: Int) =
+    (if (isLastIndex(it)) 0x00 else UVARINT_SINGLEBYTE_MAXVALUE).toByte()
```

Comment:
is varint really aware of the underlying `MutableList` we are using to store bytes? i would expect `lastByte & 0x80` to simply be an error case (since it indicates there should be more bytes, but there are not)

Reply:
it is now a non-mutable list and unless i am mistaken or I messed something up, you approved this logic [here](https://github.com/a-sit-plus/signum/pull/152).